src/v/storage/tests/compaction_e2e_test.cc

```diff
@@ -526,3 +627,554 @@ TEST_F(CompactionFixtureTest, TestReadFilledGapsWithTerms) {
+        ASSERT_EQ(consumed_kvs.size(), num_tombstones);
+
+        // Every record should be a tombstone without a value.
+        for (const auto& kv : consumed_kvs) {
+            ASSERT_EQ(kv.val, "");
+        }
+    }
+
+    // Sleep for tombstone.retention.ms time, so that the next time we attempt
+    // to compact the tombstone records will be eligible for deletion.
+    ss::sleep(tombstone_retention_ms).get();
+
+    // Generate one record, so that sliding window compaction can occur.
+    generate_data(1, 200, 1, 1, 100).get();
+    did_compact = do_sliding_window_compact(
+                    log->segments().back()->offsets().get_base_offset(),
+                    tombstone_retention_ms)
+                    .get();
+    ASSERT_TRUE(did_compact);
+    {
+        tests::kafka_consume_transport consumer(make_kafka_client().get());
+        consumer.start().get();
+        auto consumed_kvs = consumer
+                              .consume_from_partition(
+                                topic_name,
+                                model::partition_id(0),
+                                model::offset(0))
+                              .get();
+
+        // Assert there is no tombstone record left.
+        auto find_tombstone_record = [](const auto& kv) {
+            return kv.is_tombstone;
+        };
+
+        auto tombstone_it = std::find_if(
+          consumed_kvs.begin(), consumed_kvs.end(), find_tombstone_record);
+
+        // The tombstones should have been removed after second round of
+        // compaction post tombstone.retention.ms.
+        ASSERT_EQ(tombstone_it, consumed_kvs.end());
+
+        // Assert that there was only one k-v record left in the log (the one
+        // generated to allow for sliding window compaction to occur).
+        ASSERT_EQ(consumed_kvs.size(), 1);
+        ASSERT_EQ(consumed_kvs[0].key, "key100");
+        ASSERT_EQ(consumed_kvs[0].val, "val100");
+    }
+}
+
+INSTANTIATE_TEST_SUITE_P(
+  NumSegments,
+  CompactionFixtureTombstonesParamTest,
+  ::testing::Values(1, 10, 100));
+
+struct TombstonesRandomArgs {
+    static TombstonesRandomArgs create() {
+        static constexpr size_t max_segments = 100;
+        static constexpr size_t max_records = 1000;
+        static constexpr size_t max_cardinality = max_records;
+        static constexpr size_t max_batches_per_segment = 5;
+        return TombstonesRandomArgs{
+          .num_segments = random_generators::get_int(size_t{1}, max_segments),
+          .total_records = random_generators::get_int(size_t{1}, max_records),
+          .cardinality = random_generators::get_int(size_t{1}, max_cardinality),
+          .batches_per_segment = random_generators::get_int(
+            size_t{1}, max_batches_per_segment)};
+    }
+
+    size_t num_segments;
+    size_t total_records;
+    size_t cardinality;
+    size_t batches_per_segment;
+};
+
+class CompactionFixtureTombstonesRandomParamTest
```

Comment:
If you haven't already, given it seems like there's some timing involved in these tests, it'd be good to run with --gtest_repeat and --gtest_break_on_failure a few hundred/thousand times to make sure they're not flaky

Reply:
I repeated these tests many times and the timing seems to be correct, I haven't seen any flakes in my extensive testing. Running more locally now to be sure.